lm-simulator/lm_simulator/crud.py

```diff
@@ -1,129 +1,155 @@
 from typing import List
 
-from sqlalchemy import delete, select
+from buzz import enforce_defined, handle_errors, require_condition
+from fastapi import HTTPException, status
+from sqlalchemy import select
 from sqlalchemy.orm import Session
 
 from lm_simulator.models import License, LicenseInUse
 from lm_simulator.schemas import LicenseCreate, LicenseInUseCreate, LicenseInUseRow, LicenseRow
 
 
-class LicenseNotFound(Exception):
-    """The requested license is not found."""
-
+async def add_license(session: Session, license: LicenseCreate) -> LicenseRow:
+    """
+    Add a new License to the database.
+    """
+    db_license = License(**license.model_dump())
 
-class NotEnoughLicenses(Exception):
-    """The number of requested licenses is bigger than the available."""
+    with handle_errors(
+        "Can't create License, check the input data",
+        raise_exc_class=HTTPException,
+        exc_builder=lambda exc_class, msg: exc_class(
+            status_code=status.HTTP_400_BAD_REQUEST,
+            detail="Can't create License, check the input data",
+        ),
+    ):
+        session.add(db_license)
+        await session.flush()
+        await session.refresh(db_license)
 
+        return LicenseRow.model_validate(db_license, from_attributes=True)
 
-def get_licenses(session: Session) -> List[LicenseRow]:
-    db_licenses = session.execute(select(License)).scalars().all()
-    return [LicenseRow.from_orm(license) for license in db_licenses]
 
+async def list_licenses(session: Session) -> List[LicenseRow]:
+    """
+    List all Licenses in the database.
+    """
+    query = await session.execute(select(License))
+    db_licenses = query.scalars().all()
+    return [LicenseRow.model_validate(license) for license in db_licenses]
 
-def get_licenses_in_use(session: Session) -> List[LicenseInUseRow]:
-    db_licenses_in_use = session.execute(select(LicenseInUse)).scalars().all()
-    return [LicenseInUseRow.from_orm(license) for license in db_licenses_in_use]
 
+async def remove_license(session: Session, license_name: str):
+    """
+    Remove the License from the database.
 
-def get_licenses_in_use_from_name(session: Session, license_name: str) -> List[LicenseInUse]:
-    db_licenses_in_use = (
-        session.execute(select(LicenseInUse).where(LicenseInUse.license_name == license_name)).scalars().all()
+    Raises LicenseNotFound if the License does not exist.
+    """
+    query = await session.execute(select(License).where(License.name == license_name))
+    db_license = query.scalar_one_or_none()
+    enforce_defined(
+        db_license,
+        "License not found",
+        raise_exc_class=HTTPException,
+        exc_builder=lambda exc_class, msg: exc_class(
+            status_code=status.HTTP_404_NOT_FOUND,
+            detail="License not found",
+        ),
     )
```

Comment:
You could combine these into a single expression since `enforce_defined()` returns the value it checks:
```suggestion
    db_license = enforce_defined(
        query.scalar_one_or_none()
        "License not found",
        raise_exc_class=HTTPException,
        exc_builder=lambda exc_class, msg: exc_class(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="License not found",
        ),
    )
```

Reply:
Nice, I updated it!